### PR TITLE
Transport type cast information in constructorArgs to constructor selection (#1023)

### LINKED ISSUE #1023: Possibility to disambiguate constructors when creating spies
Given you have a class with two constructors, one taking a `String` argument, one taking a `Pattern` argument.
Currently you cannot create a `Spy` with `null` as argument as far as I have seen.
The only way I found is to use Objenesis to create the spy and then set any (private) fields the constructor would have set manually.
In normal code you could do `new MyClass(null as String)` (which through Spock also does not work properly as described in #1022), but `Spy(constructorArgs: [null as String])` does not work.
The cast is not carried on to where the constructor to invoke is searched and you get an ambiguous warning.

So I'd say we need either some typed-null, that can be used to find the right constructor and is then supplied as `null` to the constructor like `Spy(constructorArgs: [new Null(String)])` or a second property where you can give the types like `Spy(constructorArgs: [null], constructorTypes: [String])`.

### PR DESCRIPTION
**Pull request by Vampire**
Closes #1023

---
This change is [Reviewable](https://reviewable.io/reviews/spockframework/spock/1123)

### CODE REVIEW

File: spock-core/src/main/java/org/spockframework/compiler/AbstractDeepBlockRewriter.java [hunk 1]
```diff
@@ -22,6 +22,7 @@
 import org.codehaus.groovy.ast.stmt.*;
 
 public class AbstractDeepBlockRewriter extends StatementReplacingVisitorSupport {
+  private final AstNodeCache nodeCache;
   protected Block block;
```
Comment (szpak) [inline spock-core/src/main/java/org/spockframework/compiler/AbstractDeepBlockRewriter.java]: Could `block` be also made `final` along the way?

File: spock-core/src/main/java/org/spockframework/compiler/SpecialMethodCall.java [hunk 2]
```diff
@@ -190,10 +193,58 @@ public static SpecialMethodCall parse(MethodCallExpression methodCallExpr, @Null
...
+      .map(MapExpression.class::cast)
+      // get their entries that have a casted constructor arg
+      .flatMap(mapArgument -> mapArgument
+        .getMapEntryExpressions()
+        .stream()
+        .filter(SpecialMethodCall::hasCastedConstructorArg)
+      )
+      // wrap the casted constructor arg for example
+      // to transport "null as String" to the runtime properly
+      .forEach(constructorArgsEntry -> constructorArgsEntry
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/SpecialMethodCall.java]: Might be better to just extract this to a separate method

File: spock-core/src/main/java/org/spockframework/compiler/SpecialMethodCall.java [hunk 3]
```diff
@@ -190,10 +193,58 @@ public static SpecialMethodCall parse(MethodCallExpression methodCallExpr, @Null
...
   }
 
+  private static void wrapCastedConstructorArgs(List<Expression> arguments, AstNodeCache nodeCache) {
+    arguments
+      .stream()
+      // find map arguments
+      .filter(MapExpression.class::isInstance)
+      .map(MapExpression.class::cast)
+      // get their entries that have a casted constructor arg
+      .flatMap(mapArgument -> mapArgument
```
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/compiler/SpecialMethodCall.java]: extract to method `entriesWithCastedConstructorArg`

File: spock-core/src/main/java/org/spockframework/compiler/AbstractDeepBlockRewriter.java (hunk 1 above)
Comment (Vampire) [reply spock-core/src/main/java/org/spockframework/compiler/AbstractDeepBlockRewriter.java]: I guess not? It is reassigned in first line of `visit(Block block)`

Comment (szpak) [reply spock-core/src/main/java/org/spockframework/compiler/AbstractDeepBlockRewriter.java]: Ok, in the GH review view not everything is visible.

File: docs/interaction_based_testing.adoc [hunk 4]
```diff
@@ -792,10 +792,20 @@ SubscriberImpl subscriber = Spy(constructorArgs: ["Fred"])
 
 A spy is always based on a real object. Hence you must provide a class type rather
 than an interface type, along with any constructor arguments for the type.
-If no constructor arguments are provided, the type's default constructor will be used.
+If no constructor arguments are provided, the type's no-arg constructor will be used.
+
+If the given constructor arguments lead to an ambiguity, you can cast the constructor
+arguments as usual using `as`. If the testee for example has one constructor with a
```
Comment (leonard84) [inline docs/interaction_based_testing.adoc]: both `x as String` and `(String) x` should work with the new code right? Both variants should be mentioned

File: docs/interaction_based_testing.adoc [hunk 5]
```diff
@@ -792,10 +792,20 @@ SubscriberImpl subscriber = Spy(constructorArgs: ["Fred"])
...
+`String` parameter and one with a `Pattern` parameter and you want `null` as
+`constructorArg`:
+
+[source,groovy]
+----
+SubscriberImpl subscriber = Spy(constructorArgs: [null as String])
```
Comment (leonard84) [inline docs/interaction_based_testing.adoc]: We want to get to fully executable documentation, so every new valid code example should be included from a test. In this case probably one in the `org.spockframework.docs.interaction` namespace.

File: docs/interaction_based_testing.adoc (hunk 4 above)
Comment (Vampire) [reply docs/interaction_based_testing.adoc]: Yes, those are treated the same by the parser. I changed the tests to test both variants and the docs to mention both.

**Pull request merged**